File: .github/scripts/consensus.py
#!/usr/bin/env python3
"""Consensus aggregator for VSDD CI pipeline.

This module provides the exit-code computation for the consensus job.
The pure function decide_exit_code determines the process exit code based
on the aggregated verdict and the count of newly opened gap issues.
"""


def decide_exit_code(verdict: str, opened_count: int) -> int:
    """Compute the process exit code from verdict and opened_count.

    Args:
        verdict: Aggregated PR-scope verdict, must be "pass" or "fail".
        opened_count: Number of new (non-duplicate) gap issues filed.

    Returns:
        0 if verdict is "pass" and opened_count is 0, otherwise 1.

    Raises:
        ValueError: If verdict is not "pass" or "fail", or opened_count < 0.
    """
    if verdict not in ("pass", "fail"):
        raise ValueError(f"verdict must be 'pass' or 'fail', got {verdict!r}")
    if opened_count < 0:
        raise ValueError(f"opened_count must be >= 0, got {opened_count}")
    if verdict == "fail" or opened_count > 0:
        return 1
    return 0

File: .github/scripts/vsdd/ci-meta/consensus.ts
// ci-meta consensus exit-code logic. Pure function — given the aggregated
// PR-scope verdict and the count of newly opened gap issues, decide the
// process exit code.
//
// Replaces .github/scripts/consensus.py — ported to TS+Deno.

export type Verdict = "pass" | "fail";

/** Compute the process exit code from `verdict` and `opened`. Returns 0
 *  iff the verdict is `pass` AND no new gap issues were filed; 1
 *  otherwise. Throws on invalid input rather than silently routing
 *  malformed signals to a green exit. */
export function exit(verdict: Verdict, opened: number): 0 | 1 {
  if (verdict !== "pass" && verdict !== "fail") {
    throw new Error(`verdict must be 'pass' or 'fail', got ${JSON.stringify(verdict)}`);
  }
  if (!Number.isInteger(opened) || opened < 0) {
    throw new Error(`opened must be a non-negative integer, got ${opened}`);
  }
  if (verdict === "fail" || opened > 0) return 1;
  return 0;
}
